sklearn/_fast_gradient_boosting/grower.py

```diff
@@ -0,0 +1,481 @@
+
+    def __init__(self, depth, sample_indices, sum_gradients,
+                 sum_hessians, parent=None):
+        self.depth = depth
+        self.sample_indices = sample_indices
+        self.n_samples = sample_indices.shape[0]
+        self.sum_gradients = sum_gradients
+        self.sum_hessians = sum_hessians
+        self.parent = parent
+
+    def __repr__(self):
+        # To help with debugging
+        out = "TreeNode: depth={}, ".format(self.depth)
+        out += "samples={}".format(len(self.sample_indices))
+        if self.split_info is not None:
+            out += ", feature_idx={}".format(self.split_info.feature_idx)
+            out += ", bin_idx={}".format(self.split_info.bin_idx)
+        return out
+
+    def __lt__(self, other_node):
+        """Comparison for priority queue.
+
+        Nodes with high gain are higher priority than nodes with low gain.
+
+        heapq.heappush only need the '<' operator.
+        heapq.heappop take the smallest item first (smaller is higher
+        priority).
+
+        Parameters
+        -----------
+        other_node : TreeNode
+            The node to compare with.
+        """
+        if self.split_info is None or other_node.split_info is None:
+            raise ValueError("Cannot compare nodes with split_info")
+        return self.split_info.gain > other_node.split_info.gain
+
+
+class TreeGrower:
+    """Tree grower class used to build a tree.
+
+    The tree is fitted to predict the values of a Newton-Raphson step. The
+    splits are considered in a best-first fashion, and the quality of a
+    split is defined in splitting._split_gain.
+
+    Parameters
+    ----------
+    X_binned : array-like of int, shape=(n_samples, n_features)
+        The binned input samples. Must be Fortran-aligned.
+    gradients : array-like, shape=(n_samples,)
+        The gradients of each training sample. Those are the gradients of the
+        loss w.r.t the predictions, evaluated at iteration ``i - 1``.
+    hessians : array-like, shape=(n_samples,)
+        The hessians of each training sample. Those are the hessians of the
+        loss w.r.t the predictions, evaluated at iteration ``i - 1``.
+    max_leaf_nodes : int or None, optional(default=None)
+        The maximum number of leaves for each tree. If None, there is no
+        maximum limit.
+    max_depth : int or None, optional(default=None)
+        The maximum depth of each tree. The depth of a tree is the number of
+        nodes to go from the root to the deepest leaf.
+    min_samples_leaf : int, optional(default=20)
+        The minimum number of samples per leaf.
+    min_gain_to_split : float, optional(default=0.)
+        The minimum gain needed to split a node. Splits with lower gain will
+        be ignored.
+    max_bins : int, optional(default=256)
+        The maximum number of bins. Used to define the shape of the
+        histograms.
+    n_bins_per_feature : array-like of int or int, optional(default=None)
```

Comment:
We can't (this was the source of a bug that took me days to find lol).

`max_bins` is what the users wants as the maximum number of bins.
Since some features maybe have less than `max_bins` unique values, they will end up with an actual number of bins that is < `max_bins`. This is what we store in `n_bins_per_feature`.

Basically this test always passes:
```py
for feature_idx in range(X.shape[1]):
	assert n_bins_per_features[feature_idx] <= max_bins
```

Would it be clearer if I named it `actual_n_bins`?

Reply:
aah I see. So this is something you call, not the user. I read it with assuming the user would use it, but seems like you're calling it and you know what they mean. Yeah `actual_n_bins` would help.